src/hotspot/os/windows/os_windows.cpp

```diff
@@ -1456,27 +1456,117 @@ static int _print_module(const char* fname, address base_address,
+                                                   source_str, flag_source_str_is_null_terminated,
+                                                   *dest_utf16_str, utf16_chars_count_estimated);
+  assert(utf16_chars_count_real == utf16_chars_count_estimated, "length already checked above");
+
+  return ERROR_SUCCESS;
+}
+
+// Converts a string in the "platform" encoding to UTF16.
+static errno_t convert_to_UTF16(char const* platform_str, LPWSTR* utf16_str) {
+  return convert_to_UTF16(platform_str, CP_ACP, utf16_str);
+}
+
+static errno_t convert_UTF8_to_UTF16(char const* utf8_str, LPWSTR* utf16_str) {
+  return convert_to_UTF16(utf8_str, CP_UTF8, utf16_str);
+}
+
+// Converts a wide-character string in UTF-16 encoding to the 8-bit "platform" encoding.
+// Unless the platform encoding is UTF-8, not all characters in the source string can be represented in the dest string.
+// The function succeeds in this case anyway and just replaces these with a certain character.
+static errno_t convert_UTF16_to_platform(LPWSTR source_utf16_str, char*& dest_str) {
+  const int flag_source_str_is_null_terminated = -1;
+  const int flag_estimate_chars_count = 0;
+  int chars_count_estimated = WideCharToMultiByte(CP_ACP,
+                                                  0,
+                                                  source_utf16_str, flag_source_str_is_null_terminated,
+                                                  NULL, flag_estimate_chars_count, NULL, NULL);
+  if (chars_count_estimated == 0) {
+    dest_str = NULL;
+    return EINVAL;
+  }
+
+  dest_str = NEW_C_HEAP_ARRAY(CHAR, chars_count_estimated, mtInternal);
+
+  int chars_count_real = WideCharToMultiByte(CP_ACP,
+                                             0,
+                                             source_utf16_str, flag_source_str_is_null_terminated,
+                                             dest_str, chars_count_estimated, NULL, NULL);
+  assert(chars_count_real == chars_count_estimated, "length already checked above");
+
+  return ERROR_SUCCESS;
+}
+
+class MemoryReleaserW : public StackObj {
+private:
+  WCHAR* _object_ptr;
+
+public:
+  MemoryReleaserW(WCHAR * object_ptr) : _object_ptr(object_ptr) {}
+  ~MemoryReleaserW() { if (_object_ptr != NULL) FREE_C_HEAP_ARRAY(WCHAR, _object_ptr); }
+};
+
+class MemoryReleaser : public StackObj {
+private:
+  CHAR* _object_ptr;
+
+public:
+  MemoryReleaser(CHAR * object_ptr) : _object_ptr(object_ptr) {}
+  ~MemoryReleaser() { if (_object_ptr != NULL) FREE_C_HEAP_ARRAY(CHAR, _object_ptr); }
+};
+
 // Loads .dll/.so and
 // in case of error it checks if .dll/.so was built for the
 // same architecture as Hotspot is running on
-void * os::dll_load(const char *name, char *ebuf, int ebuflen) {
-  log_info(os)("attempting shared library load of %s", name);
+void * os::dll_load(const char *utf8_name, char *ebuf, int ebuflen) {
+  LPWSTR utf16_name = NULL;
+  errno_t err = convert_UTF8_to_UTF16(utf8_name, &utf16_name);
```

Comment:
Do you have any figures on the cost of this additional conversion in relation to startup time?

I'm already concerned to see that we have to perform each conversion twice via MultiByteToWideChar/WideCharToMultiByte, once to get the size and then to actually get the characters! This seems potentially very inefficient.

Reply:
I measured time spend converting the library path name relative to the overall time of a (successful) `os::dll_load()` call. It varies between a fraction of a percent to ~3% (see below for the actual data from a `release` build). I'll defer to your expertise wrt to these numbers.

What can be done here (without changing os::dll_load() API) is to have a "fast path" for ascii-only path names, in which case no conversion is necessary, and a "slow path" for all the rest. This will complicate things a bit, but not by much, I believe. I am certainly willing to give that a try. Let me know what do you think.


```
./build/windows-x86_64-server-release/images/jdk/bin/java -jar ./build/windows-x86_64-server-fastdebug/images/jdk/demo/jfc/SwingSet2/SwingSet2.jar
0.050% for C:\cygwin64\home\maxim\work\OpenJDK\jdk\build\windows-x86_64-server-release\images\jdk\bin\jimage.dll
2.273% for C:\cygwin64\home\maxim\work\OpenJDK\jdk\build\windows-x86_64-server-release\images\jdk\bin\java.dll
0.177% for C:\cygwin64\home\maxim\work\OpenJDK\jdk\build\windows-x86_64-server-release\images\jdk\bin\net.dll
0.541% for C:\cygwin64\home\maxim\work\OpenJDK\jdk\build\windows-x86_64-server-release\images\jdk\bin\nio.dll
0.359% for C:\cygwin64\home\maxim\work\OpenJDK\jdk\build\windows-x86_64-server-release\images\jdk\bin\zip.dll
3.186% for C:\cygwin64\home\maxim\work\OpenJDK\jdk\build\windows-x86_64-server-release\images\jdk\bin\jimage.dll
0.075% for C:\cygwin64\home\maxim\work\OpenJDK\jdk\build\windows-x86_64-server-release\images\jdk\bin\awt.dll
0.232% for C:\cygwin64\home\maxim\work\OpenJDK\jdk\build\windows-x86_64-server-release\images\jdk\bin\freetype.dll
0.136% for C:\cygwin64\home\maxim\work\OpenJDK\jdk\build\windows-x86_64-server-release\images\jdk\bin\fontmanager.dll
0.170% for C:\cygwin64\home\maxim\work\OpenJDK\jdk\build\windows-x86_64-server-release\images\jdk\bin\javajpeg.dll
```

Reply:
The core-libs folks have the experience/expertise with these character encoding issues so I will defer to them. But someone should run this through our startup benchmarks to see if it makes a difference.

Thanks,
David

Reply:
> The core-libs folks have the experience/expertise with these character encoding issues so I will defer to them.

Naoto has agreed to look at this.